h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstTfIdf.java

```diff
@@ -0,0 +1,153 @@
+package water.rapids.ast.prims.advmath;
+
+import hex.tfidf.DocumentFrequency;
+import hex.tfidf.InverseDocumentFrequency;
+import hex.tfidf.TermFrequency;
+import hex.tfidf.TfIdfPreprocessor;
+import water.Key;
+import water.MRTask;
+import water.fvec.*;
+import water.rapids.Env;
+import water.rapids.Merge;
+import water.rapids.Val;
+import water.rapids.ast.AstPrimitive;
+import water.rapids.ast.AstRoot;
+import water.rapids.vals.ValFrame;
+
+/**
+ * Primitive AST operation to compute TF-IDF values for given document corpus.
+ */
+public class AstTfIdf extends AstPrimitive<AstTfIdf> {
+
+    /**
+     * Name to be used for a column containing Inverse Document Frequency values in the output frame of this operation.
+     */
+    private static final String IDF_COL_NAME = "IDF";
+    /**
+     * Name to be used for a column containing TF-IDF values in the output frame of this operation.
+     */
+    private static final String TF_IDF_COL_NAME = "TF-IDF";
+    /**
+     * Column names to be used for preprocessed frame.
+     */
+    private static final String[] PREPROCESSED_FRAME_COL_NAMES = new String[] { "DocID", "Word" };
+
+    @Override
+    public int nargs() {
+        return 1 + 2; // (tf-idf input_frame_name preprocess)
+    }
+
+    @Override
+    public String[] args() {
+        return new String[]{ "frame", "preprocess" };
+    }
+
+    @Override
+    public Val apply(Env env, Env.StackHelp stk, AstRoot[] asts) {
+        Frame inputFrame = stk.track(asts[1].exec(env).getFrame());
+        boolean preprocess = asts[2].exec(env).getBool();
+
+        // Pre-processing
+        Frame wordFrame;
+        long documentsCnt;
+        if (preprocess) {
+            byte[] outputTypes = new byte[]{ Vec.T_NUM, Vec.T_STR };
+
+            TfIdfPreprocessor preprocessor = new TfIdfPreprocessor();
+            wordFrame = preprocessor.doAll(outputTypes, inputFrame).outputFrame(PREPROCESSED_FRAME_COL_NAMES, null);
+
+            documentsCnt = inputFrame.numRows();
+        } else {
+            wordFrame = inputFrame;
+            // TODO: Avoid converting to categorical Vec
+            documentsCnt = inputFrame.vec(0).toCategoricalVec().cardinality();
+        }
+
+        // DF
+        DocumentFrequency dfTaks = new DocumentFrequency();
```

Comment:
unnecessary variable

Reply:
Now solved by your previous suggestion (to make it static).